pyro/distributions/sine_skewed.py

```diff
@@ -0,0 +1,101 @@
+from math import pi
+
+import torch
+from torch.distributions import Uniform
+
+from pyro.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+
+
+class SineSkewed(TorchDistribution):
+    """The Sine Skewed distribution is a distribution for breaking pointwise-symmetry on a base-distributions over
+    the d-dimensional torus.
+
+    This distribution requires a base distribution on the torus. The parameter skewness can be inferred using
+    :class:`~pyro.infer.HMC` or :class:`~pyro.infer.NUTS`. The following will produce a uniform prior
+    over skewness,::
+
+        def model(data, batch_shape, event_shape):
+            n = torch.prod(torch.tensor(event_shape), -1, dtype=torch.int)
+            skewness = torch.empty((*batch_shape, n)).view(-1, n)
+            tots = torch.zeros(batch_shape).view(-1)
+            for i in range(n):
+                skewness[..., i] = pyro.sample(f'skew{i}', Uniform(0., 1 - tots))
+                tots += skewness[..., i]
+            sign = pyro.sample('sign', Uniform(0., torch.ones(skewness.shape)).to_event(len(skewness.shape)))
+            skewness = torch.where(sign < .5, -skewness, skewness)
+
+            if (*batch_shape, *event_shape) == tuple():
+                skewness = skewness.reshape((*batch_shape, *event_shape))
+            else:
+                skewness = skewness.view(*batch_shape, *event_shape)
+
+    .. note:: The base-distribution must be over a arbitrary dim torus.
+
+    .. note:: ``skewness.abs().sum() <= 1.`` and ``(skewness.abs() <= 1).all()``.
+
+    ** References: **
+      1. Sine-skewed toroidal distributions and their application in protein bioinformatics
+         Ameijeiras-Alonso, J., Ley, C. (2019)
+
+    :param base_density: base density on the d-dimensional torus; event_shape must be [..., 2] where
+        ``prod(event_shape[:-1]) == d``.
+    :param skewness: skewness of the distribution; must have same shape as base_density.event_shape, all values
+        must be in [-1,1] and ``abs(skewness).sum() <= 1``.
```

Comment:
The last requirement is not compatible with pyro tensor senmatics. The condition should not involve batch dimensions.

Reply:
Hmm, I'm not sure I follow; the last condition (`abs(skewness).sum() <= 1`) should be changed to
``` python
assert torch.all(skewness.view((torch.prod(torch.tensor(base_dist.batch_shape), dim=0), -1)).sum(-1) <= 1.)
```
Is this what you mean?

Reply:
It seems like that (though it is a bit tricky to follow those shape manipulations). To be more clarified, I guess you'll need to add some assertation to the `__init__` method:
+ assert base_dist.batch_shape broadcastable with skewness batch_shape
+ assert base_dist.event_shape is valid
+ assert skewness event_shape is valid and compatible with base_dist.event_shape
+ make sure that `arg_constraints["skewness"]` has a correct event dimension. If the event dimension is not known, you can use `arg_constraints = {'skewness': constraints.dependent}`
+ assert skewness satisfies the condition (abs(skewness).sum(over event dimension?) <= 1) for all items in batch dimensions
+ if event_dim is not decided until the construction, you can use
```
    @constraints.dependent_property
    def support(self):
        return constraints.independent(constraints.real, self.event_dim)
```
+ mention the support of base_dist, is there any restriction?...
+ ...
+ add some comments to point to the corresponding formula in the reference

I guess the main thing that makes the implementation complicated is: this class supports arbitrary event_dim base distribution. I'm not sure why that's the case. If you have some applications that require arbitrary event_dim, then adding the above assertations might be enough. Otherwise, I would suggest to stick with a specific `event_dim=1` or `event_dim=2`,... My first impression is you need `event_dim=2` where `event_shape = (d, 2)` but I could be wrong...

Reply:
Thanks for taking the time to help me; I'll make a [separate branch](https://github.com/aleatory-science/pyro/tree/feature/mulit_dim_sine_skewed) for the d-torus (I may have an application modeling peptide side-chains) and update this branch for the fixed `event_dim=2` as you suggest. If the application works out, I'll make open a new PR with the arbitrary dimension version.